src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -649,30 +660,100 @@ private class Emit(
-        (cl, None, { cont: F => f(filterCont(cont, _, _)) })
+        emitArrayIterator(a).copy(length = None).wrapContinuation(filterCont)
 
-      case x@ArrayMap(a, name, body, _) =>
+      case x@ArrayFlatMap(a, name, body) =>
         val elementTypeInfoA = coerce[Any](typeToTypeInfo(coerce[TArray](a.typ).elementType))
         val xmv = mb.newBit()
         val xvv = fb.newLocal(name)(elementTypeInfoA)
         val bodyenv = env.bind(name -> (elementTypeInfoA, xmv, xvv))
+        val bodyIt = emitArrayIterator(body, bodyenv)
+
+        val bodyCont = { (cont: F, m: Code[Boolean], v: Code[_]) =>
+          val bodyArray = bodyIt.arrayEmitter(cont)
+          Code(
+            xmv := m,
+            xmv.mux(
+              Code._empty,
+              Code(
+                xvv := v,
+                bodyArray.setup,
+                bodyArray.m.map(_.mux(
+                  Code._empty,
+                  Code(bodyIt.calcLength, bodyArray.addElements))).getOrElse(Code(bodyIt.calcLength, bodyArray.addElements)))))
+        }
+
+        emitArrayIterator(a).copy(length = None).wrapContinuation(bodyCont)
+
+      case x@ArrayMap(a, name, body, _) =>
+        val elt = coerce[TArray](a.typ).elementType
+        val elementTypeInfoA = coerce[Any](typeToTypeInfo(elt))
+        val xmv = mb.newBit()
+        val xvv = fb.newLocal(name)(elementTypeInfoA)
+        val bodyenv = env.bind(name -> (elementTypeInfoA, xmv, xvv))
         val (dobody, mbody, vbody) = emit(body, bodyenv)
         val mapCont = { (continuation: F, m: Code[Boolean], v: Code[_]) =>
           Code(
             xmv := m,
-            xvv := v,
+            xmv.mux(
+              xvv := defaultValue(elt),
+              xvv := v),
             dobody,
             continuation(mbody, vbody))
         }
-        val (cl, l, f) = emitArrayIterator(a)
-        (cl, l, { cont: F => f(mapCont(cont, _, _)) })
+        emitArrayIterator(a).wrapContinuation(mapCont)
+
+      case MakeArray(args, _) =>
+        val f = { cont: F =>
+          EmitArrayTriplet(Code._empty[Unit], None, coerce[Unit](Code(
+            for (elt <- args) yield {
+              val (doe, me, ve) = emit(elt)
+              Code(doe, cont(me, ve))
+            })))
+        }
+        ArrayIteratorTriplet(Code._empty, Some(const(args.length)), f)
+
+      case If(cond, cnsq, altr, typ) =>
+        val (docond, mcond, vcond) = emit(cond)
+        val xmcond = mb.newBit()
+        val xvcond = mb.newBit()
+        val mout = mb.newBit()
+        val aCnsq = emitArrayIterator(cnsq)
+        val aAltr = emitArrayIterator(altr)
+
+        val f = { cont: F =>
+          val addCnsq = aCnsq.arrayEmitter(cont)
+          val addAltr = aAltr.arrayEmitter(cont)
+          val setup = Code(
+            docond,
+            xmcond := mcond,
+            xmcond.mux(
+              mout := true,
+              Code(xvcond := coerce[Boolean](vcond),
+                addCnsq.setup, addAltr.setup)))
```

Comment:
Doesn't `mout` need to be initialized in the second branch, and `xvcond` initialized in the first?